File: umjunsik-lang-python/umjunsik.py
import sys

class Umjunsik:
    def __init__(self):
        self.data = [0]*256

    def toNumber(self, code):
        tokens = code.split(' ')
        result = 1
        for token in tokens:
            num = (self.data[token.count('어') - 1] if token.count('어') else 0) + token.count('.') - token.count(',')
            result *= num
        return result

    @staticmethod
    def type(code):
        if '동탄' in code:
            return 'IF'
        if '준' in code:
            return 'MOVE'
        if '화이팅!' in code:
            return 'END'
        if '식' in code and '?' in code:
            return 'INPUT'
        if '식' in code and '!' in code:
            return 'PRINT'
        if '식' in code and 'ㅋ' in code:
            return 'PRINTASCII'
        if '엄' in code:
            return 'DEF'

    def compileLine(self, code):
        if code == '':
            return None
        TYPE = self.type(code)
        
        if TYPE == 'DEF':
            var, cmd = code.split('엄')
            self.data[var.count('어')] = self.toNumber(cmd)
        elif TYPE == 'END':
            print(self.toNumber(code.split('화이팅!')[1]), end='')
            sys.exit()
        elif TYPE == 'INPUT':
            self.data[code.replace('식?', '').count('어')] = int(input())
        elif TYPE == 'PRINT':
            print(self.toNumber(code[1:-1]), end='')
        elif TYPE == 'PRINTASCII':
            value = self.toNumber(code[1:-1])
            print(chr(value) if value else '\n', end='')
        elif TYPE == 'IF':
            cond, cmd = code.replace('동탄', '').split('?')
            if self.toNumber(cond) == 0:
                return cmd
        elif TYPE == 'MOVE':
            return self.toNumber(code.replace('준', ''))

    def compile(self, code, check=True, errors=100000):
        jun = False
        recode = ''
        spliter = '\n' if '\n' in code else '~'
        code = code.rstrip().split(spliter)
        if check and (code[0].replace(" ","") != '어떻게' or code[-1] != '이 사람이름이냐ㅋㅋ' or not code[0].startswith('어떻게')):
            raise SyntaxError('어떻게 이게 엄랭이냐!')
        index = 0
        error = 0
        while index < len(code):
            errorline = index
            c = code[index].strip()
            res = self.compileLine(c)
            if jun:
                jun = False
                code[index] = recode                
            if isinstance(res, int):
                index = res-2
            if isinstance(res, str):
                recode = code[index]
                code[index] = res
                index -= 1
                jun = True

            index += 1
            error += 1
            if error == errors:
                raise RecursionError(str(errorline+1) + '번째 줄에서 무한 루프가 감지되었습니다.')

    def compilePath(self, path):
        with open(path) as file:
            code = ''.join(file.readlines())
            self.compile(code)

File: umjunsik-lang-swift/Sources/Umjunsik.swift
//
//  Umjunsik.swift
//
//
//  Created by Soongyu Kwon on 9/20/23.
//

import Foundation

class Umjunsik {
    var data: [Int]
    
    init() {
        self.data = Array(repeating: 0, count: 256)
    }
    
    func toNumber(code: String) -> Int {
        let tokens: [String] = code.components(separatedBy: " ")
        var result: Int = 1
        
        for token in tokens {
            let num = ((token.filter{ $0 == "어" }.count != 0) ? self.data[token.filter{ $0 == "어" }.count-1] : 0) + token.filter{ $0 == "." }.count - token.filter{ $0 == "," }.count
            result *= num
        }
        
        return result
    }
    
    static func type(code: String) -> String? {
        if code.contains("동탄") {
            return "IF"
        }
        if code.contains("준") {
            return "MOVE"
        }
        if code.contains("화이팅!") {
            return "END"
        }
        if code.contains("식") && code.contains("?") {
            return "INPUT"
        }
        if code.contains("식") && code.contains("!") {
            return "PRINT"
        }
        if code.contains("식") && code.contains("ㅋ") {
            return "PRINTASCII"
        }
        if code.contains("엄") {
            return "DEF"
        }
        
        return nil
    }
    
    func compileLine(code: String) -> Any? {
        if code.isEmpty {
            return nil
        }
        
        guard let TYPE = Umjunsik.type(code: code) else {
            return nil
        }

        if TYPE == "DEF" {
            let parts = code.components(separatedBy: "엄")
            let variable = parts[0]
            let cmd = parts[1]
            self.data[variable.filter{ $0 == "어" }.count] = self.toNumber(code: cmd)
        } else if TYPE == "END" {
            exit(Int32(self.toNumber(code: code.components(separatedBy: "화이팅!")[1])))
        } else if TYPE == "INPUT" {
            self.data[code.replacingOccurrences(of: "식?", with: "").filter{ $0 == "어" }.count] = Int(readLine() ?? "") ?? 0
        } else if TYPE == "PRINT" {
            print(self.toNumber(code: String(code.dropFirst().dropLast())), terminator: "")
        } else if TYPE == "PRINTASCII" {
            let value = self.toNumber(code: String(code.dropFirst().dropLast()))
            print((value != 0) ? String(UnicodeScalar(UInt8(value))) : "\n", terminator: "")
        } else if TYPE == "IF" {
            let parts = code.replacingOccurrences(of: "동탄", with: "").components(separatedBy: "?")
            let cond = parts[0]
            let cmd = parts[1]
            if self.toNumber(code: cond) == 0 {
                return cmd
            }
        } else if TYPE == "MOVE" {
            return self.toNumber(code: code.replacingOccurrences(of: "준", with: ""))
        }
        
        return nil
    }
    
    func compile(code: String, check: Bool = true, errors: Int = 100000) {
        var jun: Bool = false
        var recode: String = ""
        let splitter: String = code.contains("\n") ? "\n" : "~"
        var splittedCode = code.components(separatedBy: splitter)
        
        if check && (splittedCode[0].replacingOccurrences(of: " ", with: "") != "어떻게" || splittedCode[splittedCode.count-1] != "이 사람이름이냐ㅋㅋ" || !splittedCode[0].hasPrefix("어떻게")) {
            fatalError("어떻게 이게 엄랭이냐!")
        }
        
        var index: Int = 0
        var error: Int = 0
        while (index < splittedCode.count) {
            let errorline: Int = index
            let c: String = splittedCode[index].trimmingCharacters(in: .whitespacesAndNewlines)
            let res = self.compileLine(code: c)
            
            if jun {
                jun = false
                splittedCode[index] = recode
            }
            if res is Int {
                index = res as! Int - 2
            }
            if res is String {
                recode = splittedCode[index]
                splittedCode[index] = res as! String
                index -= 1
                jun = true
            }
            
            index += 1
            error += 1
            if error == errors {
                fatalError("\(errorline + 1)번째 줄에서 무한 루프가 감지되었습니다.")
            }
        }
    }
    
    func compilePath(path: String) {
        do {
            var code = try String(contentsOfFile: path, encoding: .utf8)
            code = code.trimmingCharacters(in: .whitespacesAndNewlines)
            self.compile(code: code)
        } catch {
            fatalError("어떻게 이게 파일이냐ㅋㅋ")
        }
    }
}
